File: sportsbook-app/betting-service/src/main/java/com/sportsbook/bettingservice/outbox/OutboxPublisher.java
package com.sportsbook.bettingservice.outbox;

import com.sportsbook.bettingservice.repo.OutboxRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Component;

import java.time.Instant;

@Component
@RequiredArgsConstructor
public class OutboxPublisher {
    private final OutboxRepository repo;
    private final OutboxSendPort sender;

    @Scheduled(fixedDelay = 500)
    public void publish() {
        var batch = repo.findTop50ByStatusOrderByCreatedAtAsc("PENDING");
        for (var e : batch) {
            try {
                sender.send(e.getTopic(), e.getKey(), e.getPayload());
                e.setStatus("SENT");
                e.setPublishedAt(Instant.now());
                repo.save(e);
            } catch (Exception ex) {
            }
        }
    }
}

File: sportsbook-app/betting-service/src/main/kotlin/com/sportsbook/bettingservice/outbox/OutboxPublisher.kt
package com.sportsbook.bettingservice.outbox

import com.sportsbook.bettingservice.repo.OutboxRepository
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Component
import java.time.Instant

@Component
class OutboxPublisher(
    private val repo: OutboxRepository,
    private val sender: OutboxSendPort,
) {

    @Scheduled(fixedDelay = 500)
    fun publish() {
        val batch = repo.findTop50ByStatusOrderByCreatedAtAsc("PENDING")
        for (event in batch) {
            try {
                val topic = event.topic ?: continue
                val key = event.key ?: continue
                val payload = event.payload ?: continue
                sender.send(topic, key, payload)
                event.status = "SENT"
                event.publishedAt = Instant.now()
                repo.save(event)
            } catch (ex: Exception) {
                // ignore and retry on next cycle
            }
        }
    }
}
